### PR TITLE
MF-522: See all button under forms widget opens the clinical views page

### PR DESCRIPTION
**Pull request by NischithJShetty**
This commit is a fix for the below ticket:
https://issues.openmrs.org/browse/MF-522

### CODE REVIEW

File: packages/esm-patient-forms-app/src/forms/forms-see-all.component.tsx [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+interface FormsProps {
+  patientUuid: string;
+  patient: fhir.Patient;
+}
+
+const Forms: React.FC<FormsProps> = ({ patientUuid, patient }) => {
+  async function setPatient(uUid) {
+    patient = (await fhir.read<fhir.Patient>({ type: 'Patient', patient: uUid })).data;
```
Comment (FlorianRappl) [inline packages/esm-patient-forms-app/src/forms/forms-see-all.component.tsx:29]: This is fishy on multiple levels.

Please rewrite using a side effect and changing a proper state.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.component.tsx [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+          form.complete = true;
+          form.lastCompleted = completeForm.lastCompleted ? completeForm.lastCompleted : null;
+        }
+      });
+      return form;
+    });
+    setFilledForms(filledForms);
+  }, [forms, completedForms]);
+
+  const RenderForm = () => {
```
Comment (FlorianRappl) [inline packages/esm-patient-forms-app/src/forms/forms-see-all.component.tsx:81]: Hm not sure why we create a component in the component -> just render directly

File: packages/esm-patient-forms-app/src/index.ts [hunk 3]
```diff
@@ -38,6 +38,8 @@ function setupOpenMRS() {
   defineConfigSchema(moduleName, {});
 
   return {
+    lifecycle: getAsyncLifecycle(() => import('./forms/forms-see-all.component'), options),
+    activate: 'forms-see-all.component',
```
Comment (FlorianRappl) [inline packages/esm-patient-forms-app/src/index.ts:42]: This is called "route" and I think the route should be more adequate for this scenario.

**Review by FlorianRappl — COMMENTED** [review]
I think we are on a great way here. Really top notch!

File: packages/esm-patient-forms-app/src/forms/forms-see-all.component.tsx (hunk 1 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.component.tsx:29]: Sure. I had adopted this from a different component but there is a need to optimise it further. I will make the necessary corrections.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.component.tsx (hunk 2 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.component.tsx:81]: Will change this as well

File: packages/esm-patient-forms-app/src/index.ts (hunk 3 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/index.ts:42]: I could have missed the topic on "route" in the 3.0 docs. It would be helpful if you could point me to the OpenMRS docs about "route". Thank you.

Comment (FlorianRappl) [reply packages/esm-patient-forms-app/src/index.ts:42]: `activate` is the old name, `route` is the new name. (See https://openmrs.github.io/openmrs-esm-core/#/main/creating_a_microfrontend - there is `load` and `route`, while you use `lifecycle` and `activate` -- these two are outdated and should be replaced with the new names)

Independent of the property name the exact route should be adjusted, too. As written, a route called "forms-see.all.component` is not good. A route should be something like `all-forms` or `forms/all` etc.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.component.tsx (hunk 1 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.component.tsx]: I have now removed the earlier function and replaced it with a 'State Hook'

File: packages/esm-patient-forms-app/src/forms/forms-see-all.component.tsx (hunk 2 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.component.tsx]: I have removed the extra component in the second commit

File: packages/esm-patient-forms-app/src/forms/form-view.component.tsx [hunk 4]
```diff
@@ -166,7 +166,7 @@ const FormView: React.FC<FormViewProps> = ({ forms, patientUuid, patient }) => {
               pageNumber={currentPage}
               totalItems={allForms.length}
               currentItems={results.length}
-              pageUrl={`$\{openmrsSpaBase}/patient/${patientUuid}/chart/`}
+              pageUrl={`./forms-see-all.component.tsx?patientUuid=${patientUuid}`}
```
Comment (FlorianRappl) [inline packages/esm-patient-forms-app/src/forms/form-view.component.tsx:169]: That is a super strange URL. Please make that a nice URL... Also make the patient UUID part of the path (i.e., a segment).

File: packages/esm-patient-forms-app/src/forms/forms-see-all.component.tsx [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  recommended = 0,
+  completed,
+  all,
+}
+
+const Forms = () => {
+  const query = new URLSearchParams(location.search);
+  const patientUuid = query.get('patientUuid');
+  const [patient, setPatient] = useState<fhir.Patient>();
+  getObjectFHIR(patientUuid).then(setPatient);
```
Comment (FlorianRappl) [inline packages/esm-patient-forms-app/src/forms/forms-see-all.component.tsx:26]: Super fishy - this will be done in each render!

Use a side-effect ...

File: packages/esm-patient-forms-app/src/forms/forms-see-all.component.tsx (hunk 1 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.component.tsx]: Optimised this further with Effect Hook

File: packages/esm-patient-forms-app/src/index.ts (hunk 3 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/index.ts]: Have made the changes accordingly, Thank you.

File: packages/esm-patient-forms-app/src/forms/form-view.component.tsx (hunk 4 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/form-view.component.tsx]: I have updated this with a new URL

File: packages/esm-patient-forms-app/src/forms/forms-see-all.component.tsx (hunk 5 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.component.tsx]: The suggested changes have been made.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+const Forms = () => {
+  const urlPathArray = window.location.pathname.split('/');
+  const pathUuid = urlPathArray[urlPathArray.length - 2];
+  const [patientUuid, setUUID] = React.useState(null);
+  const [patient, setPatient] = React.useState<fhir.Patient>();
+
+  React.useEffect(() => {
+    getObjectFHIR(patientUuid).then(setPatient);
+    setUUID(pathUuid);
```
Comment (FlorianRappl) [inline packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:30]: Why is this needed? I think the UUID is already known and does not change. Also let's not use `window.location` to read that out... `useHistory` and get the matching param instead.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+  const headerTitle = t('forms', 'Forms');
+  const [error, setError] = React.useState(null);
+  const [forms, setForms] = React.useState<Array<Form>>([]);
+  const [encounters, setEncounters] = React.useState<Array<Encounter>>([]);
+  const [completedForms, setCompletedForms] = React.useState<Array<Form>>([]);
+  const [selectedFormView, setSelectedFormView] = React.useState<FormViewState>(FormViewState.all);
+  const [filledForms, setFilledForms] = React.useState<Array<Form>>([]);
+
+  React.useEffect(() => {
+    fetchAllForms().subscribe((forms) => setForms(forms), setError);
```
Comment (FlorianRappl) [inline packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:44]: What's that with `subscribe`? Why don't we `unsubscribe`? Why no simple `openmrsFetch`?

Please avoid using these observables.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+  React.useEffect(() => {
+    const fromDate = dayjs(new Date()).startOf('day').subtract(500, 'day');
+    const toDate = dayjs(new Date()).endOf('day');
+    fetchPatientEncounters(patientUuid, fromDate.toDate(), toDate.toDate()).subscribe(
+      (encounters) => setEncounters(encounters),
+      setError,
+    );
+  }, [patientUuid]);
+
+  React.useEffect(() => {
```
Comment (FlorianRappl) [inline packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:56]: To be honest looks more like a `useMemo` to me.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    const filledForms = forms.map((form) => {
+      completedForms.map((completeForm) => {
+        if (completeForm.uuid === form.uuid) {
+          form.complete = true;
+          form.lastCompleted = completeForm.lastCompleted ? completeForm.lastCompleted : null;
+        }
+      });
+      return form;
+    });
+    setFilledForms(filledForms);
```
Comment (FlorianRappl) [inline packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:76]: Same here I think this is `useMemo`

File: packages/esm-patient-forms-app/src/index.ts [hunk 10]
```diff
@@ -38,6 +38,12 @@ function setupOpenMRS() {
   defineConfigSchema(moduleName, {});
 
   return {
+    pages: [
+      {
+        load: getAsyncLifecycle(() => import('./forms/forms-see-all'), options),
+        route: 'patient/forms',
```
Comment (FlorianRappl) [inline packages/esm-patient-forms-app/src/index.ts:44]: Miss, e.g., the patient UUID etc.

According to the hyperlink we have something like `patient/forms/${patientUuid}/forms-see-all`, but here I just see `patient/forms`. Is the `forms-see-all` not required? The UUID should be. But the URL format indicates that the segment for patient UUID is actually the form UUID.

Why not have

```
/patient/${patientUuid}/forms
```

Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/index.ts:44]: I thought even if a part (begining part) of the string specified with route matches the URL it will load the page. If it is possible to assign a value like `route: 'patient/${patientUuid}/forms'` , I will do that.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 6 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:30]: Is this `useHistory` from "react-router-dom" ? and does this involve wrapping the page with the <Router>  component? could you please point me to a script that has used `useHistory` inside the 3.0 framework?

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 7 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:44]: This file is a adaptation of [forms.component.tsx](https://github.com/openmrs/openmrs-esm-patient-chart/blob/master/packages/esm-patient-forms-app/src/forms/forms.component.tsx) and I have only made minor alterations to it to and have not tried to create the widget from scratch. Most of the code directly comes from the original file which was used to build the Form Entry widget and is now a part of the "openmrs-esm-patient-chart" repo.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 8 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:56]: This file is a adaptation of forms.component.tsx and I have only made minor alterations to it to fix the small bug and have not tried to create the widget from scratch. Most of the code directly comes from the original file which was used to build the Form Entry widget and is now a part of the "openmrs-esm-patient-chart" repo.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 9 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:76]: The explation to this is the same as above and the answer can probably be given by the original committer. Do you think we should go ahead and change all these files too? It would be interesting to find out the logic used though. Should the task of rewrting the whole widget be undertaken now or just try to sort out this minor bug for the time being is a decision to be made. Thanks.

File: packages/esm-patient-forms-app/src/forms/form-view-see-all.component.tsx [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    );
+  };
+  return (
+    <div className={styles.formContainer}>
+      <Search
+        id="searchInput"
+        labelText=""
+        className={styles.formSearchInput}
+        placeholder={t('searchForForm', 'Search for a form')}
+        onChange={(evnt) => handleSearch(evnt.target.value)}
```
Comment (FlorianRappl) [inline packages/esm-patient-forms-app/src/forms/form-view-see-all.component.tsx:126]: If you change the definition of taking `value -> foo(value)` to `evt -> foo(evt.target.value)` then there is no need for the second function; and you gain performance.

So please change the definition of `handleSearch` and pass it in here directly (i.e., `onChange={handleSearch}`)

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 6 above)
Comment (FlorianRappl) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:30]: It is from `react-router-dom`. There should be plenty of examples using `useHistory` / `react-router-dom` in this repository. Please look for it yourself (hint: there should be a search functionality in whatever IDE you are using).

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 7 above)
Comment (FlorianRappl) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:44]: We should reuse components - not copy them. This is a no go that results in barely maintainable code that is bloated and contains bugs.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 8 above)
Comment (FlorianRappl) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:56]: We should reuse components - not copy them. This is a no go that results in barely maintainable code that is bloated and contains bugs.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 9 above)
Comment (FlorianRappl) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:76]: We should reuse components - not copy them. This is a no go that results in barely maintainable code that is bloated and contains bugs.

File: packages/esm-patient-forms-app/src/index.ts (hunk 10 above)
Comment (FlorianRappl) [reply packages/esm-patient-forms-app/src/index.ts:44]: At Amazon you are also not going to a page "see-all-products" but "products". I am not sure why you don't follow general principles (i.e., things that you see around in the world, where best practices are already followed), but come up with things like that.

The routing mechanism can also use a regular expression. Have a look at the patient chart microfrontend to see how that can be done.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 7 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:44]: Exactly my point here! It was impossible to **reuse** components here. Let me explain the origins of this bug. If you go to [form-view.component.tsx](https://github.com/openmrs/openmrs-esm-patient-chart/blob/master/packages/esm-patient-forms-app/src/forms/form-view.component.tsx) and observe Ln #169, the variable "pageUrl" has been hardcoded. Then again on Ln #170 the variable "pageSize" has been **hardcoded**. This was the key component required in this situation and it was impossible re-use it. So I had to go for the closest work around. The other option is rewrting everything which is clearly beyond the scope of this ticket. Now, we need to understand why reusable components with hard coded variables are part of our repo in the first place and make corrections there.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 8 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:56]: Exactly my point here! It was impossible to reuse components here. Let me explain the origins of this bug. If you go to form-view.component.tsx and observe Ln #169, the variable "pageUrl" has been hardcoded. Then again on Ln #170 the variable "pageSize" has been hardcoded. This was the key component required in this situation and it was impossible re-use it. So I had to go for the closest work around. The other option is rewrting everything which is clearly beyond the scope of this ticket. Now, we need to understand why reusable components with hard coded variables are part of our repo in the first place and make corrections there.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 9 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:76]: As I have explained before the existing component defined in [form-view.component.tsx](https://github.com/openmrs/openmrs-esm-patient-chart/blob/master/packages/esm-patient-forms-app/src/forms/form-view.component.tsx) has hardcoded variables on Ln# 169 and Ln# 170 making reusability beyond scope. Please suggest the right solutions going forward.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 7 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:44]: But putting the old codes aside, I will rewrite certain aspects. I will make changes based on your suggestions and come up with a new commit. Thank you.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 8 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:56]: But putting the old codes aside, I will rewrite certain aspects. I will make changes based on your suggestions and come up with a new commit. Thank you.

File: packages/esm-patient-forms-app/src/index.ts (hunk 10 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/index.ts:44]: Done. Thanks!

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 6 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx]: I have implemented the other changes but in spite of a lot of attempts “useHistory” was giving errors as reported in “https://stackoverflow.com/questions/64838587/how-to-properly-use-usehistory-from-react-router-dom” so could not make use of it. A proper solution to the error could not be found anywhere, so continued with the old implementation.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 7 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx]: This is done and I have replaced the older function with "openmrsFetch"

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 8 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx]: I have made other changes but probably "useMemo" is not needed here as the fuction gets called only once. Please let me know if I am missing something.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 9 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:118]: Even this function gets invoked only once and so probably "useMemo" is not needed.

File: packages/esm-patient-forms-app/src/index.ts (hunk 10 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/index.ts]: I have changed the path format.

File: packages/esm-patient-forms-app/src/forms/form-view-see-all.component.tsx [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+};
+
+const FormView: React.FC<FormViewProps> = ({ forms, patientUuid, patient }) => {
+  const { t } = useTranslation();
+  const { currentVisit } = useVisit(patientUuid);
+  const [searchTerm, setSearchTerm] = React.useState<string>(null);
+  const [allForms, setAllForms] = React.useState<Array<Form>>(forms);
+  const { results, goTo, currentPage } = usePagination(allForms.sort(sortFormLatestFirst), 10);
+
+  const handleSearch = React.useMemo(() => debounce((searchTerm) => setSearchTerm(searchTerm), 300), []);
```
[the remark this answers is not in the document]
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/form-view-see-all.component.tsx]: I have fixed this. Thanks!

File: packages/esm-patient-forms-app/src/forms/form-view-see-all.component.tsx (hunk 11 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/form-view-see-all.component.tsx]: This is also done.

**Comment by NischithJShetty** [comment]
@FlorianRappl I have fixed most of the issues but faced some problems with one and I have mentioned that above. Please review and let me know if further changes are required. Thank you.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 8 above)
Comment (FlorianRappl) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx]: Yes you are missing that the `setCompletedForms` is causing a re-render.

Instead, use just a `useMemo`.

Your pattern (bad practice):

```js
const [state, setState] = React.useState();
React.useEffect(() => {
  // long computation giving "result"
  setState(result);
}, []);
```

Instead we can do:

```js
const state = React.useMemo(() => {
  // long computation giving "result"
  return result;
}, []);
```

much more efficient and it's clear from context what goes on...

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 9 above)
Comment (FlorianRappl) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:118]: Please read (and understand) my other comment carefully... Thanks!

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 8 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx]: Thanks for explaining. My observation was that even when `setCompletedForms` gets called enforcing a re-render the below block of code does not get executed because the dependacy array will prevent its execution:

`useEffect(() => {
    const fromDate = dayjs(new Date()).startOf('day').subtract(500, 'day');
    const toDate = dayjs(new Date()).endOf('day');
    fetchPatientEncounters(patientUuid, fromDate.toDate(), toDate.toDate()).then(setEncounters);
  }, [patientUuid]);`

I tried replacing this block with `useMemo` but it was not working probably because there is a `fetch` call happening from within this function and the useMemo fails to re-assign the value returned by the "async"  `fetch` function. But there is no problem when I use `useState` functionality. I will try implementing "useMemo" again if you want.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 9 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:118]: Even here because of the same reason above (since the value of "completedForms" is dependent upon "encounters") the returned value from async "fetch" was not get re-assigned  when useMemo was used. Please let me know if I need to try again and implement as "useMemo" itself.

**Comment by NischithJShetty** [comment]
Hi @FlorianRappl 

I have stated this as individual comments alongside the particular isssues as well. I did make a lot of attempts to replace `useState` with `useMemo` but it does not seem to be working in this situation. Even when I go to the official React docs there seems to be no mention of using `useMemo` in the place of `useState`. The docs don't really talk about using `useMemo` to initialise state variables. Please let me know if there are any docs detailing the usage of `useMemo` to initialise/ change component state.

When I changed `useState` to `useMemo` it was not updating the variable after a call to the  `openmrsFetch` function  which is `async`. I think when there is a `fetch` call being made its easier to use `useState` than `useMemo`. I could not find any reliable doc suggesting the use of `useMemo` in the place of `useState`. In spite of multiple attempts I could not make use of `useMemo` in this situation.

Sorry about this as I have tried everything possible and hope you are fine if we continue to use `useState` in this situation. 

Thank you.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 8 above)
Comment (FlorianRappl) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx]: > I tried replacing this block with useMemo but it was not working probably because there is a fetch call happening from within this function and the useMemo fails to re-assign the value returned by the "async" fetch function.

Yes, that is a good point. I did not see this beforehand, but the operation is async. I'd still say the code there can be improved (that I could not immediately see there having a resource being fetched just confirms this), but either way - let's leave it as is right now. Thanks for trying!

**Comment by FlorianRappl** [comment]
> Even when I go to the official React docs there seems to be no mention of using useMemo in the place of useState.

It is no replacement and I have not written this. Please compare again with my original example (and where the duality that has been used is the right replacement). Here, we have an async operation, which indeed is a side-effect. Ideally, however, you will reduce the number of re-renders (as you always should).

In any case let's not lose more time here.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+import styles from './forms.component.scss';
+import { EmptyState } from '@openmrs/esm-patient-common-lib';
+import { useTranslation } from 'react-i18next';
+import { toFormObject } from './forms.resource';
+import { filterAvailableAndCompletedForms } from './forms-utils';
+import { Form } from '../types';
+import EmptyFormView from './empty-form.component';
+import { getObjectFHIR } from './get-FHIR-object';
+import { openmrsFetch } from '@openmrs/esm-framework';
+import _ from 'lodash';
```
Comment (FlorianRappl) [inline packages/esm-patient-forms-app/src/forms/forms-see-all.tsx:16]: Please use `lodash-es` and a direct submodule import.

You see that already in the same module, e.g.:

```js
import first from 'lodash-es/first';
```

File: packages/esm-patient-forms-app/src/forms/pagination-see-all.component.tsx [hunk 14]
```diff
@@ -0,0 +1,52 @@
+import React from 'react';
+import Pagination from 'carbon-components-react/es/components/Pagination';
+import styles from '@openmrs/esm-patient-common-lib/src/pagination/pagination.component.scss';
+import { useTranslation } from 'react-i18next';
+import { ConfigurableLink } from '@openmrs/esm-framework';
+import { usePaginationInfo } from '@openmrs/esm-patient-common-lib/src/pagination/usePaginationInfo';
```
Comment (FlorianRappl) [inline packages/esm-patient-forms-app/src/forms/pagination-see-all.component.tsx:6]: `@openmrs/esm-patient-common-lib` is fully imported. Just use `import { ...} from '@openmrs/esm-patient-common-lib'`.

File: packages/esm-patient-forms-app/src/forms/forms-see-all.tsx (hunk 13 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/forms-see-all.tsx]: The corrections have been made.

File: packages/esm-patient-forms-app/src/forms/pagination-see-all.component.tsx (hunk 14 above)
Comment (NischithJShetty) [reply packages/esm-patient-forms-app/src/forms/pagination-see-all.component.tsx]: The corrections have been made.

**Review by FlorianRappl — APPROVED** [review]
LGTM - lets go with it!
